Motoko Intro slides (#2732)
dfinity/motoko: doc/modules/language-guide/pages/overview.adoc

== Overview

=== Motivation and Goals

A simple, useful language for the Internet Computer (IC)

* Familiar syntax
* Safe by default

> * Incorporating [line-through]*actor* _canister_ model

Reviewer: ```suggestion
* Incorporating *actor* model for canisters
```

* Seamless integration of IC features
* Making most of present and future WebAssembly

=== Key Design Points

* Object-oriented, functional & imperative

> * Objects as records of members

Reviewer: ```suggestion
* Objects as records of functions
```

* `async`/`await` for sequential programming of asynchronous messaging
* Structural typing with simple generics and subtyping

> * Safe arithmetic (unbounded and checked)

Reviewer: ```suggestion
* Safe arithmetic (both unbounded and checked)
```

* Non-nullable types by default
* Garbage collected (no manual memory management)
* JavaScript-like syntax but statically typed & sane

> Inspirations: Java(Script), C#, Swift, Pony, ML, Haskell

Reviewer: ```suggestion
Inspirations: Java, JavaScript, C#, Swift, Pony, ML, Haskell
```

=== Semantics

> * call-by-value (like Java, C, JS, ML; unlike Haskell, Nix)

Reviewer: ```suggestion
* call-by-value (like Java, C, JS, ML; unlike Haskell)
```

* declarations are locally mutually recursive
* parametric, bounded polymorphism

> * subtyping as subsumption, not coercion.

Reviewer: ```suggestion
* subtyping as zero-cost subsumption, not coercion
```

* no dynamic casts
* no inheritance

=== Implementation(s)

* implemented in OCaml (leverages `wasm` libary)
* simple reference interpreter
* less simple compiler to WebAssembly
** multipass with typed IR in each pass.
** uniform representation, unboxed arithmetic

> ** two-space gc or mark-compact gc, gc between messages (for now)

Reviewer: ```suggestion
** two-space GC or mark-compact GC, invoked after messages (for now)
```

* polymorphism by erasure

== The language

=== Expressions

* Identifiers: +
`x`, `foo_bar`, `test123`, `List`, `Map`
* Parentheses `( … )` for grouping
* Braces `{ … }` for scoping (and records)
* `;` for sequencing
* Type annotations (to help type inference): +
`(42 : Int)` +
  (zero cost)

=== Libraries

[source#impDebugInt, motoko]
....
  import Debug "mo:base/Debug";
  import Int "mo:base/Int";
....

(`import MyLib "src/MyLib"` imports a library from the local file system.)

=== Libraries

[source, motoko]
....
  import Debug "mo:base/Debug";
  import Int "mo:base/Int";
  import Trie "mo:base/Trie";

  type Users = Trie.Trie<Text, Nat>; // reference types

  Debug.print(Int.toText(7)); // reference functions/values
....

=== Blocks and declarations

> 

Reviewer: ```suggestion
```

[source.include_impDebugInt, motoko]
....
  type Delta = Nat;
  func print() {
    Debug.print(Int.toText(counter));
  };
  let d : Delta = 42;
  var counter = 1;
  counter := counter + d;
  print();
....

* Semicolon after each declaration!
* Mutually recursive
* Mutable variables marked explicitly

=== Control flow

The usual suspects...

* `do { … }`

> * `if (b) …`
> * `if (b) … else …`
> * `switch e { case (pat1) e1; …; case _ en }`
> * `while (b) …`
> * `loop …`
> * `loop …  while (b)`
> * `for (pat in e) …`
> * `return`, `return e`
> * `label l e`,  `break l e`
> * `async e`, `await e`             _(restricted)_
> * `throw`, `try … catch (x) { … }` _(restricted)_

Reviewer: ```suggestion
* `if b …`
* `if b … else …`
* `switch e { case pat1 e1; …; case _ en }`
* `while b …`
* `loop …`
* `loop …  while b`
* `for (pat in e) …`
* `return`, `return e`
* `label l e`,  `break l e`
* `async e`, `await e`             _(restricted)_
* `throw`, `try … catch x { … }` _(restricted)_
```

> 

Reviewer: Mention `do?` here?

== Primitive types

=== Unbounded integers

`Int`

`{ ..., -2, 1, 0, 1, 2, ... }`

Inferred by default for negative literals.

Literals: `13`, `0xf4`, `-20`, `+1`, `1_000_000`

=== Unbounded naturals

`Nat`

`{ 0, 1, 2, ... }`

Non-negative, trap on underflow.

Inferred by default for non-negative literals

Literals: `13`, `0xf4`, `1_000_000`

`Nat <: Int`

`Nat` is a _subtype_ of `Int`

(you can supply a `Nat` wherever an `Int` is expected)

=== Bounded numbers (trapping)

`Nat8`, `Nat16`, `Nat32`, `Nat64`, `Int8`, `Int16`, `Int32`, `Int64`

> Trap on over- and underflow. Wrap-on-trap and bit-manipulating
> operations available.

Reviewer: ```suggestion
Trap on over- and underflow; wrap-around and bit-manipulating
operations available separately
```

Needs type annotations (somewhere)

Literals: `13`, `0xf4`, `-20`, `1_000_000`

=== Floating point numbers

`Float`

IEEE 754 double precision (64 bit) semantics, normalized NaN

Inferred for fractional literals

Literals: 0, -10, `2.71`, `-0.3e+15`, `3.141_592_653_589_793_12`

=== Numeric operations

No surprises here

`- x` +
`a + b` +
`a % b` +
`a & b` +
`a << b` +
…

`a +% b, a -% b, …` for wrapping, modular arithmetic (where appropriate)

> === Characters and text

Reviewer: ```suggestion
=== Characters and Text
```

`Char`, `Text`

> Unicode! No random access.

Reviewer: ```suggestion
Unicode! Character = Unicode scalar value; no random access on text
```

....
map : <A, B>(f : A -> B, xs : [A]) -> [B]
let funcs : [<T>(T) -> T] = …
....

=== Function Declarations & Use

[source.include_impDebugInt, motoko]
....
func add(x : Int, y : Int) : Int = x + y;

func applyNTimes<T>(n : Int, x : T, f : T -> ()) {
  if (n <= 0) return;
  f(x);
  applyNTimes(n-1, x, f);
};

applyNTimes<Text>(3, "Hello!", func(x) { Debug.print(x) } );
....

* `func() { … }` short for `func() : () = { … }`
* Parametric functions
* Type instantiations may sometimes be omitted
* Anonymous functions (a.k.a. lambdas)

== Composite types

=== Tuples

`(Bool, Float, Text)`

immutable, heterogeneous, fixed size

[source#tuple, motoko]
....
let tuple = (true or false, 0.6 * 2.0, "foo" # "bar");
....

[source.include_tuple, motoko]
....
tuple.1;
....

[source.include_tuple, motoko]
....
let (_,_,t) = tuple;
t
....

=== Options

`?Text`

is either a value of that type, e.g. `?"hello"`, or `null`.

[source#display, motoko]
....
func display(x : ?Text) : Text {
  switch x {
    case (null) { "No value" };
    case (?y) { "Value: " # y };
  };
};
....

[source.include_display, motoko]
....
display(null);
....

[source.include_display, motoko]
....
display(?"Test");
....

> 

Reviewer: Simple `do?` example here?

=== Arrays (immutable)

`[Text]`

[source.include_impDebugInt, motoko]
....
let days = [ "Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun" ];

assert(days.size() == 7);

assert(days[1] == "Tue");

// days[7] will trap (fixed size)

for (d in days.vals()) { Debug.print(d) };
....

=== Arrays (mutable)

`[var Nat]`

[source, motoko]
....
let counters = [var 1, 2, 3];

assert(counters.size() == 3);

counters[1] := counters[1] + 1;

// counters[3] will trap (fixed size)

counters;
....

=== Records

> `{ first : Text; last : Text; salary : var Nat }`

Reviewer: ```suggestion
`{first : Text; last : Text; salary : var Nat}`
```

[source.include_impDebugInt, motoko]
....

> let employee = { first = "Boris"; last = "Johnson"; var salary = 81_932 };

Reviewer: ```suggestion
let employee = {first = "Boris"; last = "Johnson"; var salary = 81_932};
```

Debug.print(
  employee.first # " " # employee.last # " earns " #

>   Int.toText(employee.salary) # " pounds."

Reviewer: ```suggestion
    Int.toText(employee.salary) # " pounds."
```

);

employee.salary += 79_496;

employee;
....

=== Objects

`{ first : Text; last : Text; get : () -> Nat; add : Nat -> () }`

[source, motoko]
....
object self {
  public let first = "Boris";

>   public let last = "Johnson";

Reviewer: Seriously?

>   public let last = "Johnson";

Reviewer: Seriously?

  var salary : Nat = 81_932; // private by default
  public func get() : Nat = salary;
  public func add(bump : Nat) { salary += bump };
}
....

=== Classes

> 

Reviewer: ```suggestion
```

[source, motoko]
....
class Employee(fst : Text, lst : Text) {
  public let first = fst;
  public let last = lst;
  var salary : Nat = 0;
  public func get() : Nat = salary;
  public func add(bump : Nat) { salary += bump };
}
....

Classes are factories for constructing objects. +
A class introduces a type and a function (for constructing instances).

Just sugar for:

[source.no-repl, motoko]
....
type Employee = { first : Text; last : Text; get : () -> Nat; add : Nat -> () };

func Employee(fst : Text, lst : Text) : Employee = object { … }
....

=== Variants
`{ #Sun; #Mon; #Tue; #Wed; #Thu; #Fri; #Sat }`

[source, motoko]
....
type Day = { #Sun; #Mon; #Tue; #Wed; #Thu; #Fri; #Sat };

func toText(d : Day) : Text {
  switch d {
     case (#Sun) "Sunday";
     case (#Mon) "Monday";
     case (#Tue) "Tuesday";
     case (#Wed) "Wednesday";
     case (#Thu) "Thursday";
     case (#Fri) "Friday";
     case (#Sat) "Saturday";
   };
};

func sort(d : Day) : { #WeekDay; #WeekEnd } {
  switch d {
    case (#Sun or #Sat) #WeekEnd;  // or pattern
    case _ #WeekDay;  // wildcard pattern
  };
};
....

=== Recursive Types

[source#Lists, motoko]
....
type List = {

>   #item : { head : Text; tail : List }; // variant with payload!

Reviewer: Using a record here strikes me as rather verbose in use. I'd use:
```suggestion
  #cons : (head : Text, tail : List); // variant with payload!
```

  #empty
};

func reverse<T>(l : List<T>) : List<T> {
  func rev(l : List<T>, r : List<T>) : List<T> {
    switch l {
      case (#empty) { r };
      case (#item { head; tail }) { // a nested pattern
        rev(tail, #item {head; tail = r})
      }
    }
  };
  rev(l, #empty);
};

let s : List<Text> =
  reverse(#item { head = "A"; tail = #item { head = "B"; tail = #empty }});

let ns : List<Nat> =
  reverse(#item { head = 0; tail = #item { head = 1; tail = #empty }})
....

== Packages and modules

=== Modules

[source.no-repl, motoko]
....
// the type of base/Int.mo
module {
  type Int = Prim.Types.Int;
  toText : Int -> Text;
  abs : Int -> Nat;
  // ...
}
....

modules contain named types and values (like objects), +
but are restricted to _static_ content (pure, no state, …)

=== Module imports

[source.no-repl, motoko]
....
import Debug "mo:base/Debug";  // import from package
import Int "mo:base/Int";
import MyLib "lib/MyLib";  // import from local file MyLib.mo
....

`base` package provides basic features as separate modules.

More libraries popping up!

`MyLib.mo` _must_ contain a module or actor class, eg:

[source.no-repl, motoko]
....
module {

> 

Reviewer: ```suggestion
```

  public type List<T> = …;

  public func reverse<T>(l : List<T>) : List<T> { … };
}
....

== Platform features

=== Actor types

Like object types, but marked as `actor`:

[source#actorTypes, motoko]
....
type Broadcast = actor {
  register : Receiver -> ();
  send : Text -> async Nat;
};

type Receiver = actor {
  recv : query Text -> async Nat
};
....

_sharable_ arguments and _no_ or _async_ result type.

* `register` is a _oneway_ IC method (unawaitable).
* `send` is an IC _update_ method
* `recv` is IC _query_ method

> IC "canister" ≈ Motoko "actor"

Reviewer: ```suggestion
IC canister ≈ Motoko actor
```

=== sharable ≈ serializable

**Sharable:**

* all primitive types
* records, tuples, arrays, variants, options +
with immutable sharable components
* `actor` types
* `shared` function type

**Not sharable:**

* mutable things
* local functions
* objects (with methods)

=== A complete actor

[source, motoko]
....
import Array "mo:base/Array";

actor Broadcast {

> 

Reviewer: ```suggestion
```
